scratch_space/lib/src/scratch_space.dart

```diff
@@ -93,14 +93,13 @@ class ScratchSpace {
         if (pending != null) futures.add(pending);
       } else {
         file.createSync(recursive: true);
-        var done = () async {
-          var bytes = await reader.readAsBytes(id);
-          await _descriptorPool.withResource(() async {
-            await file.writeAsBytes(bytes);
-            // ignore: unawaited_futures
-            _pendingWrites.remove(id);
-          });
-        }();
+        var done = reader
+            .readAsBytes(id)
+            .then((bytes) =>
+                _descriptorPool.withResource(() => file.writeAsBytes(bytes)))
+            .whenComplete(() {
+          _pendingWrites.remove(id);
```

Comment:
nit: use .whenComplete(() => _pendingWrites.remove(id)) to be consistent with previous line 😀

Reply:
would need to be `..whenComplete` in that case. The `remove` would return `done` which turns this into an infinitely recursing Future

Reply:
Actually looking at the documentation, `whenComplete` doesn't return the inner future. It still returns the original future: https://api.dartlang.org/stable/1.24.3/dart-async/Future/whenComplete.html

Reply:
Oh weird indeed! When I try it that way it hangs _without_ an error just on the first compile. Perhaps `whenComplete` is not implemented as described.

Reply:
Oh of course! I totally overlooked that despite returning the result of the original future, it will only await any returned future first before doing so. Your code is correct; I was mistaken!

Reply:
ah yeah that makes sense. It ignores the result but still waits for it